File: app/timezone.go
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See License.txt for license information.

package app

import (
	"github.com/mattermost/mattermost-server/model"
	"github.com/mattermost/mattermost-server/utils"
)

func (a *App) Timezones() model.SupportedTimezones {
	if cfg := a.timezones.Load(); cfg != nil {
		return cfg.(model.SupportedTimezones)
	}
	return model.SupportedTimezones{}
}

func (a *App) LoadTimezones() {
	timezonePath := "timezones.json"

	if a.Config().TimezoneSettings.SupportedTimezonesPath != nil && len(*a.Config().TimezoneSettings.SupportedTimezonesPath) > 0 {
		timezonePath = *a.Config().TimezoneSettings.SupportedTimezonesPath
	}

	timezoneCfg := utils.LoadTimezones(timezonePath)

	a.timezones.Store(timezoneCfg)
}

File: webapp/channels/src/packages/mattermost-redux/src/selectors/entities/timezone.ts
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import timezones from 'timezones.json';

import type {UserProfile} from '@mattermost/types/users';

import {createSelector} from 'mattermost-redux/selectors/create_selector';
import {getTimezoneLabel, getUserCurrentTimezone} from 'mattermost-redux/utils/timezone_utils';

import {getCurrentUser} from './common';

export const getTimezoneForUserProfile = createSelector(
    'getTimezoneForUserProfile',
    (profile: UserProfile) => profile,
    (profile) => {
        if (profile && profile.timezone) {
            return {
                ...profile.timezone,
                useAutomaticTimezone: profile.timezone.useAutomaticTimezone === 'true',
            };
        }

        return {
            useAutomaticTimezone: true,
            automaticTimezone: '',
            manualTimezone: '',
        };
    },
);

export const getCurrentTimezoneFull = createSelector(
    'getCurrentTimezoneFull',
    getCurrentUser,
    (currentUser) => {
        return getTimezoneForUserProfile(currentUser);
    },
);

export const getCurrentTimezone = createSelector(
    'getCurrentTimezone',
    getCurrentTimezoneFull,
    (timezoneFull) => {
        return getUserCurrentTimezone(timezoneFull);
    },
);

export function generateCurrentTimezoneLabel(timezone: string) {
    if (!timezone) {
        return '';
    }

    return getTimezoneLabel(timezones, timezone);
}

export const getCurrentTimezoneLabel = createSelector(
    'getCurrentTimezoneLabel',
    getCurrentTimezone,
    generateCurrentTimezoneLabel,
);
